zkevm-circuits/src/mpt_circuit/storage_leaf/leaf_value.rs

```diff
@@ -100,890 +101,159 @@ Note that there are no checks for example for the root as the constraints to ens
-                    meta.query_advice(proof_type.is_non_existing_storage_proof, Rotation::cur());
-
-                // Note: some is_wrong_leaf constraints are in this config because
-                // leaf_non_existing config only triggers constraints for
-                // non_existing_storage proof (see q_enable).
-                constraints.push((
-                    "is_wrong_leaf needs to be 0 when not in non_existing_account proof",
-                    q_enable * (one.clone() - is_non_existing_storage_proof) * is_wrong_leaf,
-                ));
+                require!(a!(accs.key.mult) => a!(accs.key.rlc, rot_key));
+                require!(a!(value_prev) => value_rlc.prev());
+                require!(a!(value) => value_rlc);
             }
 
-            constraints
+            // Get the number of bytes used by the value
+            let value_num_bytes = matchx! {
+                is_short => 1.expr(),
+                is_long => 1.expr() + get_num_bytes_short(a!(s_main.rlp2)),
+            };
+
+            // If `is_modified_node_empty = 1`, which means an empty child, we need to
+            // ensure that the value is set to 0 in the placeholder leaf. For
+            // example when adding a new storage leaf to the trie, we have an empty child in
+            // `S` proof and non-empty in `C` proof.
+            ifx! {branch.contains_placeholder_leaf(meta, is_s) => {
+                require!(a!(s_main.rlp1) => 0);
+            }}
+
+            // Make sure the RLP encoding is correct.
+            // storage = [key, value]
+            let num_bytes = storage.num_bytes(meta);
+            // TODO(Brecht): modify the witness for empty placeholder leafs to have valid
+            // RLP encoding
+            ifx! {not!(branch.contains_placeholder_leaf(meta, is_s)) => {
+                let key_num_bytes = storage.num_bytes_on_key_row(meta);
+                require!(num_bytes => key_num_bytes.expr() + value_num_bytes.expr());
+            }};
+
+            // Check that the storage leaf is in its parent.
+            ifx! {storage.is_below_account(meta) => {
+                ifx!{storage.is_placeholder_without_branch(meta) => {
+                    // Hash of the only storage leaf which is placeholder requires empty storage root
+                    let empty_root_rlc = EMPTY_TRIE_HASH.iter().map(|v| v.expr()).collect::<Vec<_>>().rlc(&r);
+                    let storage_root_rlc = storage.storage_root_in_account_above(meta);
+                    require!(storage_root_rlc => empty_root_rlc);
+                } elsex {
+                    // Hash of the only storage leaf is storage trie root
+                    let storage_root_rlc = storage.storage_root_in_account_above(meta);
+                    require!((1, leaf_rlc, num_bytes, storage_root_rlc) => @"keccak");
+                }}
+            } elsex {
+                // TODO(Brecht): how does contains_placeholder_leaf really impact these checks?
```

Comment:
It's just that when there is a placeholder leaf, the leaf shouldn't be checked whether it's in the parent, because the leaf doesn't exist. Or did you have something else in mind?

Reply:
Yeah no that explains. I was just a bit confused that the if didn't wrap around the full parent check, but the first case (is_below_account) has its own handling for this so makes sense. Thanks!